File: spl_validator/src/analyzer/subsearch.py
"""Subsearch validation - validates [...] subsearch blocks."""
from typing import Optional
from ..parser.ast import Pipeline, Command, Subsearch
from ..models.result import ValidationResult
from ..registry import is_generating_command


def validate_subsearch(parent_cmd: Command, subsearch: Subsearch, result: ValidationResult) -> None:
    """Validate a subsearch block.
    
    Rules:
    1. Subsearch must start with a generating command
    2. Subsearch should typically end with return, table, or fields
    """
    if subsearch.pipeline is None or not subsearch.pipeline.commands:
        result.add_error(
            "SPL022",
            "Empty subsearch - subsearch must contain at least one command",
            subsearch.start,
            subsearch.end
        )
        return

    parent = parent_cmd.name.lower()
    if parent in {"appendpipe", "foreach"}:
        return
    
    # Rule 1: Must start with generating command
    first_cmd = subsearch.pipeline.commands[0]
    if not is_generating_command(first_cmd.name):
        result.add_error(
            "SPL021",
            f"Subsearch must start with a generating command (search, inputlookup, etc.), "
            f"got '{first_cmd.name}'",
            first_cmd.start,
            first_cmd.end,
            suggestion="Add 'search' at the beginning of the subsearch"
        )
    
    # Rule 2: Should end with limiting command (warning, not error)
    last_cmd = subsearch.pipeline.commands[-1]
    limiting_commands = {"return", "table", "fields", "format", "head"}
    if last_cmd.name.lower() not in limiting_commands:
        result.add_warning(
            "BEST010",
            f"Subsearch should typically end with return, table, or fields to limit output",
            last_cmd.start,
            last_cmd.end,
            suggestion=f"Consider adding '| return 100 field' or '| table field' at the end"
        )


def validate_all_subsearches(pipeline: Pipeline, result: ValidationResult) -> None:
    """Find and validate all subsearches in a pipeline."""
    for cmd in pipeline.commands:
        if cmd.subsearch:
            validate_subsearch(cmd, cmd.subsearch, result)

File: typescript/core/src/analyzer/subsearch.ts
import type { Command, Pipeline, Subsearch } from "../ast.js";
import { addError, addWarning } from "../models/result.js";
import type { ValidationResult } from "../models/result.js";
import { isGeneratingCommand } from "../registry.js";

export function validateSubsearch(
  parentCmd: Command,
  subsearch: Subsearch,
  result: ValidationResult,
): void {
  if (!subsearch.pipeline || !subsearch.pipeline.commands.length) {
    addError(
      result,
      "SPL022",
      "Empty subsearch - subsearch must contain at least one command",
      subsearch.start,
      subsearch.end,
    );
    return;
  }

  const parent = parentCmd.name.toLowerCase();
  if (parent === "appendpipe" || parent === "foreach") return;

  const firstCmd = subsearch.pipeline.commands[0]!;
  if (!isGeneratingCommand(firstCmd.name)) {
    addError(
      result,
      "SPL021",
      `Subsearch must start with a generating command (search, inputlookup, etc.), got '${firstCmd.name}'`,
      firstCmd.start,
      firstCmd.end,
      "Add 'search' at the beginning of the subsearch",
    );
  }

  const lastCmd = subsearch.pipeline.commands[subsearch.pipeline.commands.length - 1]!;
  const limiting = new Set(["return", "table", "fields", "format", "head"]);
  if (!limiting.has(lastCmd.name.toLowerCase())) {
    addWarning(
      result,
      "BEST010",
      "Subsearch should typically end with return, table, or fields to limit output",
      lastCmd.start,
      lastCmd.end,
      "Consider adding '| return 100 field' or '| table field' at the end",
    );
  }
}

export function validateAllSubsearches(pipeline: Pipeline, result: ValidationResult): void {
  for (const cmd of pipeline.commands) {
    if (cmd.subsearch) {
      validateSubsearch(cmd, cmd.subsearch, result);
    }
  }
}
